src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        /// <summary>
+        /// An implementation of <see cref="ICompilationTracker"/> that takes a compilation from another compilation tracker and updates it
+        /// to return a generated document with a specific content, regardless of what the generator actually produces. In other words, it says
+        /// "take the compilation this other thing produced, and pretend the generator gave this content, even if it wouldn't."
+        /// </summary>
+        private class GeneratedFileReplacingCompilationTracker : ICompilationTracker
+        {
+            private readonly ICompilationTracker _underlyingTracker;
+            private readonly SourceGeneratedDocumentState _replacedGeneratedDocumentState;
+
+            /// <summary>
+            /// The lazily-produced compilation that has the generated document updated. This is initialized by call to
+            /// <see cref="GetCompilationAsync"/>.
+            /// </summary>
+            [DisallowNull]
+            private Compilation? _compilationWithReplacement;
+
+            public GeneratedFileReplacingCompilationTracker(ICompilationTracker underlyingTracker, SourceGeneratedDocumentState replacementDocumentState)
+            {
+                _underlyingTracker = underlyingTracker;
+                _replacedGeneratedDocumentState = replacementDocumentState;
+            }
+
+            public bool HasCompilation => _underlyingTracker.HasCompilation;
+
+            public ProjectState ProjectState => _underlyingTracker.ProjectState;
+
+            public ICompilationTracker Clone()
+            {
+                return new GeneratedFileReplacingCompilationTracker(_underlyingTracker.Clone(), _replacedGeneratedDocumentState);
+            }
```

Comment:
why do we need to clone these...

Reply:
See https://github.com/dotnet/roslyn/pull/52094#discussion_r600895602 for the prior discussion here.